File: libs/api/src/main/java/com/example/api/ApiAdapterFactory.java
package com.example.api;

import android.support.annotation.NonNull;

import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

import io.reactivex.Observable;
import retrofit2.CallAdapter;
import retrofit2.Retrofit;

public class ApiAdapterFactory extends CallAdapter.Factory {
  @Override
  public CallAdapter<?, ?> get(@NonNull Type returnType, @NonNull Annotation[] annotations, @NonNull Retrofit retrofit) {
    if (getRawType(returnType) != Observable.class) {
      return null;
    }

    Type observableType = getParameterUpperBound(0, (ParameterizedType) returnType);
    Class<?> rawObservableType = getRawType(observableType);

    if (rawObservableType != ApiResponse.class) {
      throw new IllegalArgumentException("type must be a resource");
    }

    if (!(observableType instanceof ParameterizedType)) {
      throw new IllegalArgumentException("resource must be parametrized");
    }

    Type bodyType = getParameterUpperBound(0, (ParameterizedType) observableType);

    return new ApiCallAdapter<>(bodyType);
  }
}

File: libs/api/src/main/kotlin/com/example/api/ApiAdapterFactory.kt
package com.example.api

import io.reactivex.Observable
import retrofit2.CallAdapter
import retrofit2.Retrofit
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

class ApiAdapterFactory : CallAdapter.Factory() {
  override fun get(returnType: Type, annotations: Array<Annotation>, retrofit: Retrofit): CallAdapter<*, *>? {
    if (CallAdapter.Factory.getRawType(returnType) != Observable::class.java) {
      return null
    }

    val observableType = CallAdapter.Factory.getParameterUpperBound(0, returnType as ParameterizedType)
    val rawObservableType = CallAdapter.Factory.getRawType(observableType)

    if (rawObservableType != ApiResponse::class.java) {
      throw IllegalArgumentException("type must be a resource")
    }

    if (observableType !is ParameterizedType) {
      throw IllegalArgumentException("resource must be parametrized")
    }

    val bodyType = CallAdapter.Factory.getParameterUpperBound(0, observableType)

    return ApiCallAdapter<Any>(bodyType)
  }
}
